runtime/src/utils/buffer/write.rs

```diff
@@ -176,153 +210,97 @@ impl<B: Blob> Blob for Write<B> {
-        let inner = self.inner.read().await;
-
-        // Ensure offset read doesn't overflow
+        // Prepare `buf` to capture the read data.
         let mut buf = buf.into();
-        let data_len = buf.len();
-        let data_end = offset
-            .checked_add(data_len as u64)
+        let buf_len = buf.len(); // number of bytes to read
+
+        // Ensure the read doesn't overflow.
+        let end_offset = offset
+            .checked_add(buf_len as u64)
             .ok_or(Error::OffsetOverflow)?;
 
-        // If the data required is beyond the buffer end, return an error
-        let buffer_start = inner.position;
-        let buffer_end = buffer_start + inner.buffer.len() as u64;
+        // Acquire a read lock on the buffer.
+        let buffer = self.buffer.read().await;
 
-        // Ensure we don't read beyond the logical end of the blob
-        if data_end > buffer_end {
+        // If the data required is beyond the size of the blob, return an error.
+        if end_offset > buffer.size() {
             return Err(Error::BlobInsufficientLength);
         }
 
-        // Case 1: Read entirely from the underlying blob (before buffer)
-        if data_end <= buffer_start {
-            return inner.blob.read_at(buf, offset).await;
-        }
-
-        // Case 2: Read entirely from the buffer
-        if offset >= buffer_start {
-            let buffer_offset = (offset - buffer_start) as usize;
-            let end_offset = buffer_offset + data_len;
-
-            if end_offset > inner.buffer.len() {
-                return Err(Error::BlobInsufficientLength);
-            }
-
-            buf.put_slice(&inner.buffer[buffer_offset..end_offset]);
+        // Read any bytes from the buffer that overlap with the requested range.
+        let remaining = buffer.read(buf.as_mut(), offset);
+        if remaining == 0 {
             return Ok(buf);
         }
 
-        // Case 3: Read spans both blob and buffer
-        let blob_bytes = (buffer_start - offset) as usize;
-        let buffer_bytes = data_len - blob_bytes;
-
-        // Read from blob first
-        let blob_part = vec![0u8; blob_bytes];
-        let blob_part = inner.blob.read_at(blob_part, offset).await?;
-
-        // Copy blob data and buffer data to result
-        buf.as_mut()[..blob_bytes].copy_from_slice(blob_part.as_ref());
-        buf.as_mut()[blob_bytes..].copy_from_slice(&inner.buffer[..buffer_bytes]);
+        // If bytes remain, read directly from the blob.
+        let blob_part = self.blob.read_at(vec![0u8; remaining], offset).await?;
+        buf.as_mut()[..remaining].copy_from_slice(blob_part.as_ref());
 
         Ok(buf)
     }
 
-    async fn write_at(&self, buf: impl Into<StableBuf> + Send, offset: u64) -> Result<(), Error> {
-        // Acquire a write lock on the inner state
-        let mut inner = self.inner.write().await;
+    async fn write_at(&self, data: impl Into<StableBuf> + Send, offset: u64) -> Result<(), Error> {
```

Comment:
nit: for consistency, we should probably call this `buf` rather than `data`

Reply:
I thought it was more clear to use "data" for data that was being provided, and "buf" for buffering data that was being returned, hence that change. But can revert if you find that confusing.

Reply:
ah got it. restored.